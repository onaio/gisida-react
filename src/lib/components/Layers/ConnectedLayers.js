/**
 * Handle display of submenu and layer items by utilizing redux state.
 * Keeps track of open submenus in redux state
 * 
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Layer from '../Layer/Layer';
import { connect } from 'react-redux';
import { Actions } from 'gisida';
const mapStateToProps = (state, ownProps) => {
  const MAP = state[ownProps.mapId];
  const { mapId, layers, currentRegion, preparedLayers } = ownProps;
  return {
    openGroups: MAP.openGroups,
    mapId,
    layers,
    currentRegion,
    preparedLayers,
  };
};
export class ConnectedLayers extends Component {
  /**
   * Toggle open and close a menu group
   * @param {Object} e Event
   * @param {Object} layer Group to be toggled
   * @param {number} groupCount Group count
   */
  toggleSubMenu(e, layer, groupCount) {
    e.preventDefault();
    const { openGroups } = this.props;
    let index = -1;
    let count = 0;
    let found = false;
    while (!found && count < openGroups.length) {
      if (openGroups[count].count === groupCount && openGroups[count].group === layer) {
        index = count;
        found = true;
      }
      count += 1;
    }
    this.props.dispatch(Actions.toggleGroups(this.props.mapId, layer, index, false, groupCount));
  }
  /**
   * Get the group open state
   * @param {number} groupCount Group count
   * @param {string} groupName Group name
   * @returns {boolean} Whether the group is open or not
   */
  isGroupOpen(groupCount, groupName) {
    return (
      this.props.openGroups &&
      this.props.openGroups.filter(group => group.count === groupCount && group.group === groupName)
        .length
    );
  }
  render() {
    const { mapId, layers, currentRegion, preparedLayers, auth } = this.props;
    let layerKeys;
    let layerObj;
    let layerItem = [];
    const subLayerIds = [];
    if (!preparedLayers) {
      return false;
    }
    layerKeys = Object.keys(preparedLayers);
    for (let lo = 0; lo < layerKeys.length; lo += 1) {
      layerObj = this.props.preparedLayers[layerKeys[lo]];
      if (layerObj.layers) {
        for (let s = 0; s < layerObj.layers.length; s += 1) {
          subLayerIds.push(layerObj.layers[s]);
        }
      }
    }
    layers.forEach(layer => {
      if (
        (!currentRegion || (layer.region && layer.region === currentRegion)) &&
        !subLayerIds.includes(layer.id)
      ) {
        if (layer.id && (!auth || !auth.authConfigs)) {
          if (this.props.layerItem) {
            const CustomLayerItem = this.props.layerItem;
            layerItem.push(<CustomLayerItem key={layer.id} mapId={mapId} layer={layer} />);
          } else {
            layerItem.push(<Layer key={layer.id} mapId={mapId} layer={layer} />);
          }
        } else if (layer.id && auth) {
          const { authConfigs, userInfo } = auth;
          let activeId = layer.id;
          let users;
          if (activeId.indexOf('.json') !== -1) {
            // layer ids from github shared repo have .json suffix
            // split to remove .json part
            activeId = activeId.split('.')[0];
          }
          const LocalAuthConfig = JSON.parse(localStorage.getItem('authConfig'));
          users = authConfigs.LAYERS && authConfigs.LAYERS[activeId]; // list of users with access to the layer
          authConfigs.LAYERS = authConfigs.LAYERS || LocalAuthConfig.LAYERS;
          users = authConfigs.LAYERS[activeId]; 
          // list of users with access to the layer
          // check if logged in user exists in the list of users
          // who have access to the layer
          if (
            (users && userInfo && users.includes(userInfo.username)) ||
            (authConfigs.LAYERS &&
              authConfigs.LAYERS.ALL &&
              authConfigs.LAYERS.ALL.includes(userInfo.username))
          ) {
            if (this.props.layerItem) {
              const CustomLayerItem = this.props.layerItem;
              layerItem.push(<CustomLayerItem key={layer.id} mapId={mapId} layer={layer} />);
            } else {
              layerItem.push(<Layer key={layer.id} mapId={mapId} layer={layer} />);
            }
          }
        } else {
          Object.keys(layer).forEach((d, i) => {
            layerItem = layerItem.concat([
              <li>
                <a
                  key={`${d}-${i}-link`}
                  className="sub-category"
                  onClick={e => this.toggleSubMenu(e, d, layer[d].count)}
                >
                  {d}
                  <span
                    className={`category glyphicon glyphicon-chevron-${
                      this.isGroupOpen(layer[d].count, d) ? 'down' : 'right'
                    }`}
                  />
                </a>
              </li>,
              this.isGroupOpen(layer[d].count, d) ? (
                <ConnectedLayers
                  openGroups={this.props.openGroups}
                  dispatch={this.props.dispatch}
                  layerItem={this.props.layerItem}
                  key={`${d}-${i}`}
                  mapId={mapId}
                  layers={layer[d].layers}
                  currentRegion={currentRegion}
                  preparedLayers={preparedLayers}
                  auth={this.props.auth}
                />
              ) : null,
            ]);
          });
        }
      }
      return null;
    });
    return <ul className="layers">{layerItem}</ul>;
  }
}
ConnectedLayers.propTypes = {
  mapId: PropTypes.string.isRequired,
  layers: PropTypes.arrayOf(PropTypes.any).isRequired,
  preparedLayers: PropTypes.arrayOf(PropTypes.any),
  currentRegion: PropTypes.string,
  auth: PropTypes.object
};
export default connect(mapStateToProps)(ConnectedLayers);