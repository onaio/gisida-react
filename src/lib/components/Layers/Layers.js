import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Layer from '../Layer/Layer';

export class Layers extends Component {
  constructor(props) {
    super(props);
    const groups = {};
    if (this.props.layers) {
      this.props.layers.forEach((layer) => {
        if (!layer.id) {
          Object.keys(layer).forEach((l) => {
            groups[l] = { isOpen: false };
          });
        }
      })
    }
    this.state = groups;
  }

  toggleSubMenu(e, layer) {
    e.preventDefault();
    this.setState({
      ...this.state,
      [layer]: { isOpen: !this.state[layer].isOpen },
    });
  }

  render() {
    const { mapId, layers, currentRegion, preparedLayers } = this.props;

    let layerKeys;
    let layerObj;
    let layerItem = [];
    const subLayerIds = [];

    if (!preparedLayers) {
      return false;
    }

    layerKeys = Object.keys(preparedLayers);

    for (let lo = 0; lo < layerKeys.length; lo += 1) {
      layerObj = this.props.preparedLayers[layerKeys[lo]];
      if (layerObj.layers) {
        for (let s = 0; s < layerObj.layers.length; s += 1) {
          subLayerIds.push(layerObj.layers[s]);
        }
      }
    }

    layers.forEach((layer) => {
      if ((!currentRegion
        || (layer.region
          && layer.region === currentRegion))
        && !subLayerIds.includes(layer.id)) {
        if (layer.id) {
          layerItem.push((<Layer
            key={layer.id}
            mapId={mapId}
            layer={layer}
          />))
        } else {
          Object.keys(layer).forEach((d, i) => {
            layerItem = layerItem.concat([
              (
                <a
                  key={`${d}-${i}-link`}
                  className="sub-category"
                  onClick={(e) => this.toggleSubMenu(e, d)}
                >
                  {d}
                  <span
                    className={`category glyphicon glyphicon-chevron-${this.state[d].isOpen ? 'down' : 'right'}`}
                  />
                </a>
              ),
              (this.state[d].isOpen ?
                <Layers
                  key={`${d}-${i}`}
                  mapId={mapId}
                  layers={layer[d].layers}
                  currentRegion={currentRegion}
                  preparedLayers={preparedLayers}
                />
              : null)
            ]);
          });
        }
      }
      return null;
    });

    return (
      <ul className="layers">
        {layerItem}
      </ul>
    );
  }
}

Layers.propTypes = {
  mapId: PropTypes.string.isRequired,
  layers: PropTypes.arrayOf(PropTypes.any).isRequired,
  currentRegion: PropTypes.string,
};

export default Layers;
