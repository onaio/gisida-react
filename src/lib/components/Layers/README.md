# Layers Component

The menu category layers can be rendered using the `Layers` component or the `ConnectedLayers` component.

## Layers

`Layers` component is used to render menu category groups, nested subgroups if any and their layers.
Uses local state to keep track of open groups.

Example usage:

```
  <Layers
    mapId={mapId}
    layers={category.layers}
    currentRegion={currentRegion}
    preparedLayers={preparedLayers}
    auth={this.props.AUTH}
    />
```

## ConnectedLayers

`ConnectedLayers` is used to render menu category groups, nested subgroups if any and their layers.
Uses redux state to keep track of open groups.

This component also allows you to pass a component that renders a custom `<li>` element which will serve as a custom layer item in place of `src/components/Layer/Layers`

Example usage:

```
<ConnectedLayers
  layerItem={CustomLayerComponent}
  mapId={mapId}
  layers={category.layers}
  currentRegion={currentRegion}
  preparedLayers={preparedLayers}
  auth={this.props.AUTH}
/>
```

The usage of the component requires you to have passed a `useConnectedLayers` prop in your client project when
instantiating the `Menu`

```
<Menu useConnectedLayers/>
```

The custom layer item can also be directly passed to the `Menu` as a prop

```
<Menu layerItem={CustomLayerComponent} useConnectedLayers />
```

### Why choose ConnectedLayers over Layers?

The advantage of using this component over the `Layers` component is that different `Menu` instances will
share the same state of open groups. If a situation such as this arises, usage of this component will
suffice. 

An example of such a situation is when you would like to navigate from the map to another view
but would like to keep the map's menu in that view.

```
<Map mapId='map-1'>
  <Menu useConnectedLayers>
</Map>

<SomeOtherView>
  {/** This component will use map-1 menu **/}

  <Menu useConnectedLayers>
</SomeOtherView>
```