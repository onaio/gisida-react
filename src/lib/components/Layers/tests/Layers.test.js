import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Layers from './../Layers'
import layerObj from './fixtures/sample-layer.json';

describe('Layers', () => {
  layerObj.id = 'sample-layer';
  const layers = [layerObj];
  const preparedLayers = {'sample-layer-id': layerObj}
  
  const componentWrapper = shallow(
    <Layers
      mapId={'map-1'}
      layers={layers}
      currentRegion={undefined}
      preparedLayers={preparedLayers}
    />
  );
  it('component renders correctly', () => {
    const json = toJson(componentWrapper)
    expect(json).toMatchSnapshot();
  });
});