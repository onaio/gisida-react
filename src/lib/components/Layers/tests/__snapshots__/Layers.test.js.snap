// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layers component renders correctly 1`] = `
<ul
  className="layers"
>
  <Connect(Layer)
    key="sample-layer"
    layer={
      Object {
        "category": "Boundaries and Place Labels",
        "id": "sample-layer",
        "label": "Boundaries",
        "paint": Object {
          "line-color": "#444",
          "line-opacity": 0.6,
          "line-width": 1,
        },
        "source": Object {
          "layer": "som_adm_2-0j5bly",
          "type": "vector",
          "url": "mapbox://ona.6gl2shzi",
        },
        "type": "line",
        "visible": true,
      }
    }
    mapId="map-1"
  />
</ul>
`;
